README.org
#+html: <p align="left">
#+html:	<a href="https://github.com/xhcoding/emacs-aichat/actions/workflows/test.yml"><img src="https://github.com/xhcoding/emacs-aichat/actions/workflows/test.yml/badge.svg"/></a>
#+html:    <a href ="https://github.com/xhcoding/emacs-aichat/blob/main/README.zh-CN.org"><img src="https://img.shields.io/badge/README-%E7%AE%80%E4%BD%93%E4%B8%AD%E6%96%87-555555.svg"/></a>
#+html: </p>

* AI Chat in Emacs

Use AI chat features in Emacs, currently including Bing Chat and OpenAI’s ChatGpt.

* Installation

** Requirements

- [[https://github.com/ahyatt/emacs-websocket][emacs-websocket]] To communicate with Bing Chat, if you have already installed this package, please update to the latest version.
- [[https://github.com/chuntaro/emacs-async-await][emacs-async-await]]
- [[https://github.com/jrblevin/markdown-mode][markdown-mode]]

** Install

Clone the repository to your local machine, and add the path to your =load-path= :

#+begin_src sh
  git clone https://github.com/xhcoding/emacs-aichat.git ~/.emacs.d/site-lisp/emacs-aichat
#+end_src

#+begin_src elisp
  (add-to-list 'load-path "~/.emacs.d/site-lisp/emacs-aichat")
#+end_src

add the following code to your emacs startup file:

#+begin_src elisp
  (require 'aichat)
#+end_src

Or load separately

#+begin_src elisp
  (require 'aichat-bingai)
  ;; or
  (require 'aichat-openai)
#+end_src

* Bing Chat

#+html: <p align="center"><img src="images/aichat-bingai-chat.png"/></p>

[[https://github.com/xhcoding/emacs-aichat/blob/main/images/examples.org][View more examples]]


Below is information on setting up Bing AI. If you are interested in learning about using OpenAI, please scroll down to the OpenAI section.

** Setup

*** Check access (Required)

- Install the latest version of Microsoft Edge
- Open http://bing.com/chat
- If you see a chat feature, you are good to go

*** Getting authentication (Required)

emacs-aichat uses browser cookies to log in to Bing Chat, so you need to log in to http://www.bing.com first in your browser.

There are two ways to get the browser's cookie, you can choose any convenient method.

**** Option one

If you have a Python environment, run =pip3 install browser_cookie3= to install =browser_cookie3=, then open the terminal and execute the following code:


#+begin_src sh
  python -c "import browser_cookie3;list(map(lambda c: print('{} {} {} {} {} {}'.format(c.name, c.value, c.expires,c.domain, c.secure, c.path)), filter(lambda c: c.domain in ('.bing.com', 'www.bing.com'), browser_cookie3.edge(domain_name='bing.com'))))"
#+end_src

If you can see the output and there is a =_U= field in the output, then you can log in successfully with this method.

**** Option two

- Install the cookie editor extension for [[https://microsoftedge.microsoft.com/addons/detail/cookieeditor/neaplmfkghagebokkhpjpoebhdledlfi][Edge]]
- Go to bing.com
- Open the extension
- Click "Export" on the bottom right (This saves your cookies to clipboard)
- Paste your cookies into a file cookies.json
- Set =aichat-bingai-cookies-file= to your cookies.json path

*** Proxy [Optional]

Currently only tested http proxy, set =aichat-bingai-proxy= as proxy address, for example:

#+begin_src elisp
  (setq aichat-bingai-proxy "localhost:51837")
#+end_src

*If your IP address is in China, you must set up a proxy to access it.*

** Usage

*** Commands

- =aichat-bingai-chat= : Send a conversation message, adding =C-u= prefix will start a new conversation
- =aichat-bingai-assistant= : Send the selected region or the input content, and the returned result is displayed in the Buffer defined by =aichat-bingai-assistant-buffer=
- =aichat-bingai-replace-or-insert= : Send the selected region or the input content, and replace the selected region or insert at the current position with the returned result

*** Customized fixed questioning approach.

Use =aichat-bingai-prompt-create= to create a fixed way of asking questions, such as:

#+begin_src elisp
  (aichat-bingai-prompt-create "translator"
                               :input-prompt "Please translate: "
                               :text-format "Please identify the language I am speaking. If it is Chinese, please translate it into English; if it is not Chinese, please translate it into Chinese. Your answer must only contain the translated content. The following is the sentence I want to translate:\n%s"
                               :chat t
                               :assistant t
                               :replace-or-insert t)
#+end_src

The above configuration will generate three functions: =aichat-bingai-chat-translator=, =aichat-bingai-assistant-translator=, and =aichat-bingai-replace-or-insert-translator=, which have the same effect as the above three commands, but the content of the query will be in the format of =text-format=.


*** Options

- =aichat-bingai-cookies-file= : The path of cookies file.
- =aichat-bingai-conversation-style= : Conversation style, optional values are: =creative= , =balanced= or =precise= , default is =balanced=
- =aichat-bingai-chat-file= : a file that saves chat history
- =aichat-bingai-chat-display-function= : a function that displays chat files
- =aichat-bingai-assistant-buffer= : Display the buffer name of the content returned by =aichat-bingai-assistant=
- =aichat-bingai-assistant-display-function= : The function to display =aichat-bingai-assistant-buffer=, default to =display-buffer=

* OpenAI

#+html: <p align="center"><img src="images/aichat-openai-chat.png"/></p>

** Setup

*** AppKey

emacs-aichat obtains AppKey through =auth-source-pick-first-password=, and writes your appkey into =~/.authinfo= file, with the following format:

#+begin_src conf
  machine platform.openai.com login aichat-openai password your-app-key
#+end_src

*** Proxy [Optional]

Currently only tested http proxy, set =aichat-openai-proxy= as proxy address, for example:

#+begin_src elisp
  (setq aichat-openai-proxy "localhost:51837")
#+end_src



** Usage

*** Chat

1. Create a new chat file using =aichat-openai-chat=.
2. Enter the chat content.
3. Use =aichat-openai-chat-send-buffer (C-c C-c)= to send the entire contents of the buffer,
   or use =aichat-openai-chat-last-heading (C-c C-l)= to only send the last heading message.

*** Assistant

- =aichat-openai-assistant= : Send the selected region or the input content, and the returned result is displayed in the Buffer defined by =aichat-openai-assistant-buffer=
- =aichat-openai-replace-or-insert= : Send the selected region or the input content, and replace the selected region or insert at the current position with the returned result

*** Customized fixed questioning approach.

Use =aichat-openai-prompt-create= to create a fixed way of asking questions, such as:

#+begin_src elisp
  (aichat-openai-prompt-create "translator"
                               :input-prompt "Please translate: "
                               :text-format "Please identify the language I am speaking. If it is Chinese, please translate it into English; if it is not Chinese, please translate it into Chinese. Your answer must only contain the translated content. The following is the sentence I want to translate:\n%s"
                               :assistant t
                               :replace-or-insert t)
#+end_src

The above configuration will generate two functions: =aichat-openai-assistant-translator=, and =aichat-openai-replace-or-insert-translator=, which have the same effect as the above three commands, but the content of the query will be in the format of =text-format=.

*** Options

- =aichat-openai-proxy= : set up an HTTP proxy that requests OpenAI
- =aichat-openai-assistant-buffer= : the buffer name of the OpenAI Assistant result
- =aichat-openai-assistant-display-function= : the function of display assistant buffer
- =aichat-openai-chat-directory= : Directory for storing chat files
- =aichat-openai-chat-display-function= : Function to display chat buffer

* Feedback Issues

To feedback your issue, please follow these steps:
1. Open the debug mode by typing the command =aichat-toggle-debug=.
2. [[https://github.com/xhcoding/emacs-aichat/issues/new][Submit an issue]] and attach the content of the =*AICHAT-DEBUG*= buffer when an error occurs.


* Acknowledgements

- [[https://github.com/acheong08/EdgeGPT][EdgeGPT]]
- [[https://github.com/manateelazycat/mind-wave][mind-wave]]
- [[https://github.com/f/awesome-chatgpt-prompts][awesome-chatgpt-prompts]]


CHANGELOG.org

* Changelog

** 1.0.2 - 2023-05-05

- add =aichat-openai-chat=
- fix some bugs

** 1.0.1 - 2023-03-23

- fix some bugs
- add =aichat-bingai-prompt-create=

** 1.0.0 - 2023-03-14

🎉 Welcome to emacs-aichat v1.0.0 🎉

This is the first release of emacs-aichat, a package that lets you chat with AI in Emacs. You can have fun conversations with AI, ask questions, get answers, and even generate creative content like poems, stories, code, and more.

We hope you enjoy using emacs-aichat and have a great time chatting with AI. If you have any feedback or issues, please let me know on GitHub or Emacs China.

Happy chatting! 😊


test/server.js
const http = require("http");

let count = 0;

let server = http.createServer(function(req, res) {
    if (req.url === "/stream") {
        res.writeHead(200, {
            "Content-Type": "text/event-stream",
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "Access-Control-Allow-Origin": '*',
            "Transfer-Encoding": "chunked",
        });
        interval = setInterval(function() {
            res.write("event: test\ndata: this is data\ndata: this is new-line\n\n");
            count++;
            if (count === 3) {
                clearInterval(interval);
                res.socket.end();
                server.close();
            }
        }, 1000);

        req.socket.addListener("close", function() {
            clearInterval(interval);
        }, false);

    }
});

server.listen(12345, "127.0.0.1");


test/bingai-test.el
;;; bingai-test.el --- bingai-test.el   -*- lexical-binding: t; -*-

;; Filename: bingai-test.el
;; Description: bingai-test.el
;; Author: xhcoding <xhcoding@foxmail.com>
;; Maintainer: xhcoding <xhcoding@foxmail.com>
;; Copyright (C) 2023, xhcoding, all rights reserved.
;; Created: 2023-03-04 22:46:50
;; Version: 0.1
;; Last-Updated: 2023-03-04 22:46:50
;;           By: xhcoding
;; URL: https://github.com/xhcoding/emacs-aichat
;; Keywords: 
;; Compatibility: GNU Emacs 30.0.50
;;
;; Features that might be required by this library:
;;
;; 
;;

;;; This file is NOT part of GNU Emacs

;;; License
;;
;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 3, or (at your option)
;; any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program; see the file COPYING.  If not, write to
;; the Free Software Foundation, Inc., 51 Franklin Street, Fifth
;; Floor, Boston, MA 02110-1301, USA.

;;; Commentary:
;;
;; bingai-test.el
;; 

;;; Installation:
;;
;; Put bingai-test.el to your load-path.
;; The load-path is usually ~/elisp/.
;; It's set in your ~/.emacs like this:
;; (add-to-list 'load-path (expand-file-name "~/elisp"))
;;
;; And the following to your ~/.emacs startup file.
;;
;; (require 'bingai-test)
;;
;; No need more.

;;; Customize:
;;
;; 
;;
;; All of the above can customize by:
;;      M-x customize-group RET bingai-test RET
;;

;;; Change log:
;;
;; 2023/03/04
;;      * First released.
;;

;;; Acknowledgements:
;;
;; 
;;

;;; TODO
;;
;; 
;;

;;; Require
(require 'ert)

;;; Code:

(ert-deftest aichat-ci ()
  (should (= 1 1)))

(provide 'bingai-test)

;;; bingai-test.el ends here


aichat.el
;;; aichat.el --- aichat.el   -*- lexical-binding: t; -*-

;; Filename: aichat.el
;; Description: aichat.el
;; Author: xhcoding <xhcoding@foxmail.com>
;; Maintainer: xhcoding <xhcoding@foxmail.com>
;; Copyright (C) 2023, xhcoding, all rights reserved.
;; Created: 2023-03-12 15:06:06
;; Version: 0.1
;; Last-Updated: 2023-03-12 15:06:06
;;           By: xhcoding
;; URL: https://github.com/xhcoding/emacs-aichat
;; Keywords: 
;; Compatibility: GNU Emacs 30.0.50
;;
;; Features that might be required by this library:
;;
;; 
;;

;;; This file is NOT part of GNU Emacs

;;; License
;;
;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 3, or (at your option)
;; any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program; see the file COPYING.  If not, write to
;; the Free Software Foundation, Inc., 51 Franklin Street, Fifth
;; Floor, Boston, MA 02110-1301, USA.

;;; Commentary:
;;
;; aichat.el
;; 

;;; Installation:
;;
;; Put aichat.el to your load-path.
;; The load-path is usually ~/elisp/.
;; It's set in your ~/.emacs like this:
;; (add-to-list 'load-path (expand-file-name "~/elisp"))
;;
;; And the following to your ~/.emacs startup file.
;;
;; (require 'aichat)
;;
;; No need more.

;;; Customize:
;;
;; 
;;
;; All of the above can customize by:
;;      M-x customize-group RET aichat RET
;;

;;; Change log:
;;
;; 2023/03/12
;;      * First released.
;;

;;; Acknowledgements:
;;
;; 
;;

;;; TODO
;;
;; 
;;

;;; Require

;;; Code:

(defgroup aichat nil
  "AI Chat in Emacs."
  :group 'tools
  :prefix "aichat-")

(require 'aichat-bingai)

(require 'aichat-openai)


(provide 'aichat)

;;; aichat.el ends here


aichat-openai.el
;;; aichat-openai.el --- aichat-openai   -*- lexical-binding: t; -*-

;; Filename: aichat-openai.el
;; Description: aichat-openai
;; Author: xhcoding <xhcoding@foxmail.com>
;; Maintainer: xhcoding <xhcoding@foxmail.com>
;; Copyright (C) 2023, xhcoding, all rights reserved.
;; Created: 2023-03-10 12:34:27
;; Version: 0.1
;; Last-Updated: 2023-03-10 12:34:27
;;           By: xhcoding
;; URL: https://github.com/xhcoding/emacs-aichat
;; Keywords:
;; Compatibility: GNU Emacs 30.0.50
;;
;; Features that might be required by this library:
;;
;;
;;

;;; This file is NOT part of GNU Emacs

;;; License
;;
;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 3, or (at your option)
;; any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program; see the file COPYING.  If not, write to
;; the Free Software Foundation, Inc., 51 Franklin Street, Fifth
;; Floor, Boston, MA 02110-1301, USA.

;;; Commentary:
;;
;; aichat-openai
;;

;;; Installation:
;;
;; Put aichat-openai.el to your load-path.
;; The load-path is usually ~/elisp/.
;; It's set in your ~/.emacs like this:
;; (add-to-list 'load-path (expand-file-name "~/elisp"))
;;
;; And the following to your ~/.emacs startup file.
;;
;; (require 'aichat-openai)
;;
;; No need more.

;;; Customize:
;;
;;
;;
;; All of the above can customize by:
;;      M-x customize-group RET aichat-openai RET
;;

;;; Change log:
;;
;; 2023/03/10
;;      * First released.
;;

;;; Acknowledgements:
;;
;;
;;

;;; TODO
;;
;;
;;

;;; Require

(require 'aichat-util)

(require 'markdown-mode)

;;; Code:

(defgroup aichat-openai nil
  "OpenAI in Emacs."
  :group 'aichat
  :prefix "aichat-openai-")

(defcustom aichat-openai-proxy nil
  "Http proxy of request openai api.

(setq aichat-openai-proxy \"localhost:51837\")"
  :group 'aichat-openai
  :type 'string)

(defun aichat-openai--default-api-key-function ()
  "Fetch the API key with auth-source."
  (auth-source-pick-first-password :host "platform.openai.com"))

(defcustom aichat-openai-api-key #'aichat-openai--default-api-key-function
  "OpenAI key as a string or a function that loads and returns it."
  :group 'aichat-openai
  :type '(choice (function :tag "Function")
                 (string :tag "String")))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;; API ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defcustom aichat-openai-domain "https://api.openai.com"
  "OpenAI domain."
  :group 'aichat-openai
  :type 'string)

(defconst aichat-openai--chat-completions-url "/v1/chat/completions" "The url of chat completions api.")

(defconst aichat-openai--chat-completions-model "gpt-3.5-turbo" "The model of chat completions api.")

(defconst aichat-openai--chat-completions-request-type "POST" "The request type of chat completions api.")

(defun aichat-openai--make-http-headers ()
  "Create http headers to send in requests to the API."
  `(("Content-Type" . "application/json")
    ("Authorization" . ,(format "Bearer %s" (if (functionp aichat-openai-api-key)
                                                (funcall aichat-openai-api-key)
                                              aichat-openai-api-key)))))

(cl-defun aichat-openai-make-chat-messages (&rest settings
                                                  &key
                                                  user
                                                  system
                                                  assistant)
  "Make chat completions messages.

(aichat-openai--make-chat-messages
 :user \"hello\"
 :assistant \"Hello there, how may I assist you today?\")"
  (vconcat
   (cl-loop for (key value) on settings by #'cddr
            collect (list :role (string-trim-left (symbol-name key) ":") :content value))))

(cl-defun aichat-openai-chat-completions (messages &rest settings
                                                   &key
                                                   on-success
                                                   on-error
                                                   &allow-other-keys
                                                   )
  "Request openai chat completions api.

Look https://platform.openai.com/docs/api-reference/chat."

  (let* ((args (cl-loop for (key value) on settings by #'cddr
                        append (unless (member key '(:on-success :on-error))
                                 (list key value))))
         (data (aichat-json-serialize (append
                                       (list
                                        :model aichat-openai--chat-completions-model
                                        :messages messages)
                                       args))))

    (aichat-debug "Request %s with data:\n%s\n" aichat-openai--chat-completions-url data)

    (promise-then (aichat-http (concat aichat-openai-domain aichat-openai--chat-completions-url)
                               :proxy aichat-openai-proxy
                               :type aichat-openai--chat-completions-request-type
                               :headers (aichat-openai--make-http-headers)
                               :data data)
                  (lambda (value)
                    (aichat-debug "Received: %s" value)
                    (seq-let (status headers body) value
                      (if (string= "200" (car status))
                          (when on-success
                            (funcall on-success (aichat-json-parse body)))
                        (when on-error
                          (funcall on-error body)))))
                  (lambda (err)
                    (when on-error
                      (funcall on-error err))))))

(cl-defun aichat-openai-chat-completions-stream (messages callback &rest settings
                                                          &key
                                                          on-success
                                                          on-error
                                                          &allow-other-keys
                                                          )
  "Request openai chat completions api with stream mode.

Look https://platform.openai.com/docs/api-reference/chat for more request params."
  (let* ((args (cl-loop for (key value) on settings by #'cddr
                        append (unless (member key '(:on-success :on-error))
                                 (list key value))))
         (data (aichat-json-serialize (append
                                       (list
                                        :model aichat-openai--chat-completions-model
                                        :messages messages
                                        :stream t)
                                       args))))

    (aichat-debug "Request %s with data:\n%s\n" aichat-openai--chat-completions-url data)

    (promise-then (aichat-http-event-source (concat aichat-openai-domain aichat-openai--chat-completions-url)
                                            (lambda (event-id event-data)
                                              (aichat-debug "Received event: %s, data: \n%s\n" event-id event-data)
                                              (unless (string= event-data "[DONE]")
                                                (when callback
                                                  (funcall callback (aichat-json-parse event-data)))))
                                            :proxy aichat-openai-proxy
                                            :type aichat-openai--chat-completions-request-type
                                            :headers (aichat-openai--make-http-headers)
                                            :data data)
                  (lambda (value)
                    (seq-let (status headers body) value
                      (if (string= "200" (car status))
                          (when on-success
                            (funcall on-success body))
                        (when on-error
                          (funcall on-error body)))))
                  (lambda (err)
                    (when on-error
                      (funcall on-error err))))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;; Message API ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defun aichat-openai-chat-completions-content (msg)
  "[choices][0][message][content]"
  (aichat-json-access msg "{choices}[0]{message}{content}"))

(defun aichat-openai-chat-completions-delta-content (msg)
  "[choices][0][delta][content]"
  (aichat-json-access msg "{choices}[0]{delta}{content}"))

(defun aichat-openai-chat-completions-delta-role (msg)
  "[choices][0][delta][role]"
  (aichat-json-access msg "{choices}[0]{delta}{role}"))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;; Chat ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defcustom aichat-openai-chat-directory (expand-file-name "aichat/openai" user-emacs-directory)
  "The directory of save chat file."
  :group 'aichat-openai
  :type 'string)

(defcustom aichat-openai-chat-display-function 'switch-to-buffer
  "The function of display chat buffer."
  :group 'aichat-openai
  :type 'symbol)

(defun aichat-openai-chat--buffer-messages (buffer)
  "Make messages with BUFFER's content."
  (with-current-buffer buffer
    (save-mark-and-excursion
      (goto-char (point-min))
      (let ((messages (list))
            (role)
            (message-beg)
            (message-end)
            (message))
        (catch 'finished
          (while t
            (if (re-search-forward "^# \\([a-zA-Z]+\\)\n" nil t)
                (setq role (downcase (string-trim (match-string 1))))
              (throw 'finished nil))
            (setq message-beg (point))
            (if (re-search-forward "^#" nil t)
                (progn
                  (backward-char)
                  (setq message-end (point)))
              (setq message-end (point-max)))
            (setq message (string-trim (buffer-substring-no-properties message-beg message-end)))
            (push (intern (concat ":" role)) messages)
            (push message messages)))
        (reverse messages)))))

(defun aichat-openai-chat--heading-messages (buffer)
  "Make messages with BUFFER's last heading content."
  (with-current-buffer buffer
    (save-mark-and-excursion
      (goto-char (point-max))
      (when (re-search-backward "^# " nil t)
        (save-restriction
          (narrow-to-region (point) (point-max))
          (aichat-openai-chat--buffer-messages (current-buffer)))))))

(defun aichat-openai-chat--send-messages (messages buffer)
  "Send MESSAGES to OpenAI, insert response to BUFFER."
  (aichat-openai-chat-completions-stream
   (apply #'aichat-openai-make-chat-messages messages)
   (lambda (msg)
     (let ((delta-role (aichat-openai-chat-completions-delta-role msg))
           (delta-content (aichat-openai-chat-completions-delta-content msg)))
       (when (buffer-live-p buffer)
         (with-current-buffer buffer
           (save-mark-and-excursion
             (goto-char (point-max))
             (when delta-role
               (insert (format "\n# %s%s\n\n" (upcase (substring delta-role 0 1)) (substring delta-role 1))))
             (when delta-content
               (insert delta-content)))))))
   :on-success (lambda (_)
                 (when (buffer-live-p buffer)
                   (with-current-buffer buffer
                     (goto-char (point-max))
                     (insert "\n\n# User\n\n"))))
   :on-error (lambda (err)
               (message "error: %s"err))))

(defun aichat-openai-chat-send-buffer ()
  "Send current buffer content to OpenAI."
  (interactive)
  (when-let* ((buffer (current-buffer))
              (messages (aichat-openai-chat--buffer-messages buffer)))
    (aichat-openai-chat--send-messages messages buffer)))

(defun aichat-openai-chat-send-last-heading ()
  "Send last heading content to OpenAI."
  (interactive)
  (when-let* ((buffer (current-buffer))
              (messages (aichat-openai-chat--heading-messages buffer)))
    (aichat-openai-chat--send-messages messages buffer)))

;;;###autoload
(add-to-list 'auto-mode-alist '("\\.aichat\\'" . aichat-openai-chat-mode))

;;;###autoload
(define-derived-mode aichat-openai-chat-mode markdown-mode "aichat-openai-chat"
  "Major mode for openai chat."
  (setq-local markdown-hide-markup t)
  (when (string-empty-p (buffer-string))
    (insert "# User\n"))
  (define-key aichat-openai-chat-mode-map (kbd "C-c C-c") #'aichat-openai-chat-send-buffer)
  (define-key aichat-openai-chat-mode-map (kbd "C-c C-l") #'aichat-openai-chat-send-last-heading))

;;;###autoload
(defun aichat-openai-chat (buffer)
  "If the chat buffer is not found, create a new chat buffer, otherwise switch to the opened chat buffer."
  (interactive (list (let ((opened-buffers (cl-loop for buffer in (buffer-list)
                                                    append (with-current-buffer buffer
                                                             (when (derived-mode-p 'aichat-openai-chat-mode)
                                                               (list (buffer-name)))))))
                       (if (and opened-buffers (not (and (car current-prefix-arg)
                                                         (= (car current-prefix-arg) 4))))
                           (if (= 1 (length opened-buffers))
                               (car opened-buffers)
                             (completing-read "Select buffer: " opened-buffers))
                         (when-let* ((name
                                      (completing-read "Chat name: "
                                                       (if (file-exists-p aichat-openai-chat-directory)
                                                           (directory-files aichat-openai-chat-directory nil
                                                                            "^\\([^.]\\|\\.[^.]\\|\\.\\..\\)")
                                                         (list))))
                                     (filename (expand-file-name (if (string-suffix-p ".aichat" name) name
                                                                   (concat name ".aichat"))
                                                                 aichat-openai-chat-directory)))
                           (make-directory aichat-openai-chat-directory t)
                           (find-file-noselect filename))))))
  (funcall aichat-openai-chat-display-function buffer))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;; Assistant ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defcustom aichat-openai-assistant-buffer "*Aichat-OpenAI-Assistant*"
  "The buffer of show assistant message."
  :group 'aichat
  :type 'string)

(defcustom aichat-openai-assistant-display-function 'display-buffer
  "The function of display `aichat-openai-assistant-buffer'."
  :group 'aichat
  :type 'symbol)

(defun aichat-openai-assistant-get-buffer ()
  (get-buffer-create aichat-openai-assistant-buffer))

(defun aichat-openai-assistant (text)
  "Send the region or input to Bing and display the returned result to `aichat-openai-assistant-buffer'."
  (interactive (list (aichat-read-region-or-input "Input text: ")))
  (aichat-openai-chat-completions (aichat-openai-make-chat-messages
                                   :user text)
                                  :on-success (lambda (msg)
                                                (let ((content (aichat-openai-chat-completions-content msg))
                                                      (buffer (aichat-openai-assistant-get-buffer)))
                                                  (with-current-buffer buffer
                                                    (goto-char (point-max))
                                                    (insert content)
                                                    (insert "\n\n"))
                                                  (funcall aichat-openai-assistant-display-function buffer)))
                                  :on-error (lambda (err)
                                              (message "error: %s"err))))

(defun aichat-openai-replace-or-insert (text)
  "Send the region or input to OpenAI and replace the selected region or insert at the current position with the returned result."
  (interactive (list (aichat-read-region-or-input "Input text: ")))
  (when (and text (not (string-empty-p text)))
    (let* ((cur-buf (current-buffer))
           (cur-pos (with-current-buffer cur-buf (point)))
           (reg-beg (when (use-region-p) (region-beginning)))
           (reg-end (when (use-region-p) (region-end))))
      (aichat-openai-chat-completions (aichat-openai-make-chat-messages
                                       :user text)
                                      :on-success (lambda (msg)
                                                    (when-let ((content (aichat-openai-chat-completions-content msg)))
                                                      (with-current-buffer cur-buf
                                                        (if (and reg-beg reg-end)
                                                            (replace-region-contents reg-beg reg-end (lambda () content))
                                                          (goto-char cur-pos)
                                                          (insert content)))))
                                      :on-error (lambda (err)
                                                  (message "Error: %s" err))))))

(cl-defmacro aichat-openai-prompt-create (name &rest args
                                               &key
                                               (input-prompt "Input text: ")
                                               (text-format "%s")
                                               (assistant nil)
                                               (replace-or-insert nil))
  "This macro will generate two functions: aichat-openai-assistant-name or aichat-openai-replace-or-insert-name.

INPUT-PROMPT: The prompt before the user input in minibuffer.
TEXT-FORMAT: Formating string, %s is replaced by what the user input."
  (let ((assistant-func (intern (format "aichat-openai-assistant-%s" name)))
        (replace-func (intern (format "aichat-openai-replace-or-insert-%s" name))))
    `(progn
       (when ,assistant
         (defun ,assistant-func(text)
           (interactive (list (aichat-read-region-or-input ,input-prompt)))
           (when text
             (let ((query (format ,text-format text)))
               (aichat-openai-assistant query)))))
       (when ,replace-or-insert
         (defun ,replace-func(text)
           (interactive (list (aichat-read-region-or-input ,input-prompt)))
           (when text
             (let ((query (format ,text-format text)))
               (aichat-openai-replace-or-insert query))))))))

(provide 'aichat-openai)

;;; aichat-openai.el ends here
